_posts/2018-04-09-synching-sia.md

```diff
@@ -0,0 +1,131 @@
+---
+title: Syncing Sia in 73 minutes
+layout: post
+date: "2018-04-09"
+description: "An experiment to see how fast I can perform the initial Sia blockchain sync"
+tags:
+- sia
+- blockchain
+- benchmark
+permalink: "/sia-sync-73-min/"
+---
+
+*This is a guest post by Sia community contributor [Thomas Bennett](https://github.com/tbenz9).*
+
+I can sync Sia from a block height of 0 to 149000 in 73 minutes.  Here’s how I do it.
+
+## Zero subscribers
+
+As mentioned in my previous article [Benchmarking Sia’s Blockchain Sync](https://blog.spaceduck.io/sia-blockchain-sync/) Sia spends an average of 55 minutes or 19.1% of its time notifying subscribers.  I submitted the pull request [#2809](https://github.com/NebulousLabs/Sia/pull/2809) so Sia only notifies subscribers if there are any.  By running Sia with only the `consensus` and `gateway` modules loaded there will be zero subscribers.
+
+My `siad` command with minimal modules loaded.
+
+```shell
+./siad -M cg -d /root/Sia/
+```
+
+With zero subscribers Sia skips the notify subscribers functions and shaves 55 minutes off the synchronization time.  Note that with only the `consensus` and `gateway` modules loaded you cannot rent, host, or use the Sia wallet.
+
+## Reduce RPC Timeouts
+
+Sia has long timeouts when waiting for blocks to download.  I’ve submitted the pull request [#2922](https://github.com/NebulousLabs/Sia/pull/2922) to reduce some of the long timeouts.  In order to sync in 73 minutes, I dropped the timeouts in [syncronize.go](https://github.com/NebulousLabs/Sia/blob/master/modules/consensus/synchronize.go) **WAY** down.  I first tried setting `relayHeaderTimeout`, `sendBlkTimeout` and `sendBlocksTimeout` to 2 seconds, this was too low, I was timing out on all my peers and couldn’t download any blocks.  After bumping `sendBlkTimeout` and `sendBlocksTimeout` to 10 seconds I was able to stay connected to most peers and download the blocks very quickly.  Any peers that could not deliver my batch of 10 blocks in 10 seconds or less was disconnected and Sia would look for new peers.
```

Comment:
>I first tried setting `relayHeaderTimeout`, `sendBlkTimeout` and `sendBlocksTimeout` to 2 seconds

We need to introduce these first. Most readers aren't going to be familiar enough with the Sia codebase to know what these functions are. They should be able to understand the blog post without clicking to any external links.

If we can explain these timeouts in prose rather than by variable name, that would be ideal. Like in plain English, what are these events Sia is waiting on?

Also, in general I recommend linking by commit ID rather than branch. Commit IDs stay forever, whereas if they rewrite the file or rename it, a link to the branch will break:

https://github.com/NebulousLabs/Sia/blob/49704703740e12aa59abda5606346faac411f0b4/modules/consensus/synchronize.go